File: internal/trace/trace.go
package trace

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/Hiyajomaho-num9/opencode-go-codex/internal/responses"
)

type Trace struct {
	Dir string
	ID  string
	mu  sync.Mutex
}

func New(root string) (*Trace, error) {
	if root == "" {
		return nil, nil
	}
	id := time.Now().Format("20060102-150405-") + strings.TrimPrefix(responses.NewID(""), "_")[:8]
	dir := filepath.Join(root, id)
	if err := os.MkdirAll(dir, 0o700); err != nil {
		return nil, err
	}
	return &Trace{Dir: dir, ID: id}, nil
}

func (t *Trace) WriteJSON(name string, value any, redact bool) {
	if t == nil {
		return
	}
	if redact {
		value = Redact(value)
	}
	b, err := json.Marshal(value)
	if err != nil {
		b = []byte(`null`)
	}
	_ = os.WriteFile(filepath.Join(t.Dir, name), append(b, '\n'), 0o600)
}

func (t *Trace) WriteText(name string, data []byte) {
	if t == nil {
		return
	}
	_ = os.WriteFile(filepath.Join(t.Dir, name), data, 0o600)
}

func (t *Trace) AppendText(name string, data []byte) {
	if t == nil {
		return
	}
	t.mu.Lock()
	defer t.mu.Unlock()
	f, err := os.OpenFile(filepath.Join(t.Dir, name), os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o600)
	if err != nil {
		return
	}
	defer f.Close()
	_, _ = f.Write(data)
}

func (t *Trace) AppendEvent(event string, data any) {
	if t == nil {
		return
	}
	b, err := json.Marshal(map[string]any{"event": event, "data": data})
	if err != nil {
		return
	}
	t.AppendText("response_events.jsonl", append(b, '\n'))
}

func Redact(value any) any {
	switch v := value.(type) {
	case map[string]any:
		out := map[string]any{}
		for key, child := range v {
			lower := strings.ToLower(key)
			if strings.Contains(lower, "token") || strings.Contains(lower, "key") || strings.Contains(lower, "secret") || strings.Contains(lower, "password") || strings.Contains(lower, "authorization") {
				out[key] = "<redacted>"
			} else {
				out[key] = Redact(child)
			}
		}
		return out
	case []any:
		out := make([]any, len(v))
		for i, child := range v {
			out[i] = Redact(child)
		}
		return out
	default:
		return v
	}
}

func Meta(t *Trace, data map[string]any) {
	if t == nil {
		return
	}
	data["trace_id"] = t.ID
	t.WriteJSON("meta.json", data, true)
}

func Error(t *Trace, err error) {
	if t != nil && err != nil {
		t.WriteText("error.txt", []byte(fmt.Sprintf("%v\n", err)))
	}
}

File: adapter/trace.py
import os
import time
import uuid
from pathlib import Path

from .jsonutil import json_dumps


SENSITIVE_KEYS = {
    "api_key",
    "apikey",
    "authorization",
    "auth",
    "key",
    "password",
    "secret",
    "token",
}


def make_trace(root, enabled=True):
    if not root or not enabled:
        return None
    trace_id = time.strftime("%Y%m%d-%H%M%S-") + uuid.uuid4().hex[:8]
    trace_dir = Path(root).expanduser() / trace_id
    trace_dir.mkdir(parents=True, exist_ok=True)
    return Trace(trace_dir, trace_id)


def redact(value):
    if isinstance(value, dict):
        redacted = {}
        for key, child in value.items():
            lower = str(key).lower()
            if any(marker in lower for marker in SENSITIVE_KEYS):
                redacted[key] = "<redacted>"
            else:
                redacted[key] = redact(child)
        return redacted
    if isinstance(value, list):
        return [redact(item) for item in value]
    return value


class Trace:
    def __init__(self, path, trace_id):
        self.path = path
        self.trace_id = trace_id
        self._events_file = None

    def write_json(self, name, data, redact_secrets=True):
        payload = redact(data) if redact_secrets else data
        (self.path / name).write_text(json_dumps(payload) + "\n", encoding="utf-8")

    def write_text(self, name, data):
        if isinstance(data, bytes):
            data = data.decode("utf-8", "replace")
        (self.path / name).write_text(str(data), encoding="utf-8")

    def append_response_event(self, event, data):
        if self._events_file is None:
            self._events_file = (self.path / "response_events.jsonl").open("a", encoding="utf-8")
        self._events_file.write(json_dumps({"event": event, "data": data}) + "\n")
        self._events_file.flush()

    def close(self):
        if self._events_file is not None:
            self._events_file.close()
            self._events_file = None


def write_meta(trace, **data):
    if trace:
        trace.write_json("meta.json", {"trace_id": trace.trace_id, **data})
